### PR TITLE
FIX: prevent holiday duplicates when username is changed

### PR DESCRIPTION
**Pull request by lis2**
When username was changed, the scheduled job was creating duplicated holidays.

In addition to fix, migration was added to remove duplicated and incorrect records.

### CODE REVIEW

File: spec/jobs/scheduled/create_holiday_events_spec.rb [hunk 1]
```diff
@@ -111,6 +111,17 @@
     expect(CalendarEvent.pluck(:region, :description, :start_date, :username)).to eq([])
   end
 
+  it "does not create duplicates when username is changed" do
+    frenchy
+    DiscourseCalendar::CreateHolidayEvents.new.execute(nil)
+    frenchy.update!(username: "new_username")
```
Comment (martin-brennan) [inline spec/jobs/scheduled/create_holiday_events_spec.rb:117]: I think we just need another assertion here:

```ruby
created_event = CalendarEvent.last
expect(created_event.username).to eq(frenchy.username)
```

File: spec/jobs/scheduled/create_holiday_events_spec.rb [hunk 2]
```diff
@@ -111,6 +111,17 @@
...
+
+    expect { DiscourseCalendar::CreateHolidayEvents.new.execute(nil) }.not_to change {
+      CalendarEvent.count
+    }
+    expect(CalendarEvent.last.username).to eq("new_username")
```
Comment (martin-brennan) [inline spec/jobs/scheduled/create_holiday_events_spec.rb:122]: Then this change:

```suggestion
    expect(created_event.reload.username).to eq("new_username")
```

File: db/migrate/20231123233308_delete_duplicated_holidays.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+class DeleteDuplicatedHolidays < ActiveRecord::Migration[7.0]
+  def up
+    execute <<~SQL
+      DELETE
+        FROM calendar_events ce
+      WHERE
+        ce.id IN (SELECT ce2.id FROM calendar_events ce2
+                  INNER JOIN users ON users.id = ce2.user_id
```
Comment (martin-brennan) [inline db/migrate/20231123233308_delete_duplicated_holidays.rb:10]: Is it possible for a different user to create the event for someone else?

File: db/migrate/20231123233308_delete_duplicated_holidays.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+                  WHERE ce2.post_id IS NULL
+                    AND ce2.username != users.username)
```
Comment (martin-brennan) [inline db/migrate/20231123233308_delete_duplicated_holidays.rb:12]: Do we have to worry about case insensitivity here?

File: db/migrate/20231123233308_delete_duplicated_holidays.rb (hunk 3 above)
Comment (lis2) [reply db/migrate/20231123233308_delete_duplicated_holidays.rb:10]: No, Calendar events without post id are holidays only generated by scheduled job

File: db/migrate/20231123233308_delete_duplicated_holidays.rb (hunk 4 above)
Comment (lis2) [reply db/migrate/20231123233308_delete_duplicated_holidays.rb:12]: I think it is ok. Right now, if someone change username from `kris` to `Kris`, we would have duplicated holiday. We still want to delete `kris` and display properly `Kris`

**Review by martin-brennan — APPROVED** [review]

**Pull request merged**